### ISSUE TITLE
provide an Icinga/Nagios check

### ISSUE THREAD

**Issue by calestyo** [liske/needrestart#32]
Hey.

Would be nice, if there was a Icinga/Nagios check for needrestart, ideally part of monitoring-plugins.org.

Such check would obviously report the number of processes/daemons in need of a restart (of course also as perfdata).

I think it would give back the following values:
- whether a kernel restart is required
- total number of processes to be restarted
- total number of services to be restarted
- total number of user processes to be restarted (this should really be from the UID range, that the respective distro reserves for normal users[0])
- total number of root user processes to be restarted
- total number of non-root/non-normal-user processes to be restarted (again, see e.g. [0] for Debian)
- perhaps in addition, the total number of processes that needrestart, with it's current settings, would restart by it's own, if $nrconf{defno} is not set to 1,... to give an overview of what's actually "necessary".

I think it should further allow via options, to specify, which of the above values (when >0) should lead to a check status of WARNING and which to a status of ERROR.

If the check fails, or needrestart isn't isn't found, UNKNOWN may be the state of choice.

Well, have a look at https://www.monitoring-plugins.org/doc/guidelines.html especially for the perfdata and multiline output format.

Cheers,
Chris.

[0] In Debian this would be specified here: https://www.debian.org/doc/debian-policy/ch-opersys.html#s9.2.2

**Comment by liske** [comment]
There is already a plugin mode available:
```
# needrestart -p; echo "RC=$?"
CRIT - Kernel: 4.3.3-tl1+, Services: 6 (!), Containers: none, Sessions: 2 (!)|Kernel=0;0;;0;2 Services=6;;0;0 Containers=0;;0;0 Sessions=2;0;;0
RC=2
```

It already provides performance data, too. The severity is currently hard coded[1] but I'm going to make it configurable with one of the next releases.

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=827023

**Comment by calestyo** [comment]
Ah.. I see O:-D

Nice... well perhaps you can still pick some ideas from my text above,... i.e. more numbers, especially totals vs. what would needrestart do per default,... and also the splitting to UID regions (root, system user, normal user, etc. pp.)

I also just thought about whether it would make sense to provide a wrapper check_needrestart script and maybe alreday a command definition in /etc/nagios-plugins/config/ ... ?

Cheers,
Chris.

**Comment by liske** [comment]
Hi,

example files for using *needrstart* with *nagios* have been added  long time ago (see 1680da87f484d5cd43c95c522ca7277d2acc6327). They are not shipped in the offical Debian packages, feel free to file a Debian wishlist bug (maybe to provide a *monitoring-plugins-needrestart* package).

I've improved the plugin with a detailed multiline output and configurable services status (depending on outdated kernel, services, sessions, containers).
